### PR TITLE
refactor: migrate gateway implemenation to component arch

### PR DESCRIPTION
**Pull request by zdtsw**
## Description

final part of the origin PR https://github.com/opendatahub-io/opendatahub-operator/pull/2539
- add new SecretNamespace in gatewayconfig
- merge to action files in gateway into one
~~- make 2 replica for auth proxy deployment~~

ref https://issues.redhat.com/browse/RHOAIENG-34952

## Merge criteria

- [ ] You have read the [contributors guide](https://github.com/opendatahub-io/opendatahub-operator/blob/incubation/CONTRIBUTING.md).
- [ ] Commit messages are meaningful - have a clear and concise summary and detailed explanation of what was changed and why.
- [ ] Pull Request contains a description of the solution, a link to the JIRA issue, and to any dependent or related Pull Request.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work
- [ ] The developer has run the integration test pipeline and verified that it passed successfully

### E2E test suite update requirement

When bringing new changes to the operator code, such changes are by default required to be accompanied by extending and/or updating the E2E test suite accordingly.

To opt-out of this requirement:
1. **Please inspect the [opt-out guidelines](https://github.com/opendatahub-io/opendatahub-operator/blob/main/docs/e2e-update-requirement-guidelines.md)**, to determine if the nature of the PR changes allows for skipping this requirement
2. If opt-out is applicable, provide justification in the dedicated `E2E update requirement opt-out justification` section below
3. Check the checkbox below:
- [ ] Skip requirement to update E2E test suite for this PR
4. Submit/save these changes to the PR description. This will automatically trigger the check.

## Summary by CodeRabbit

* **New Features**
  * Optional OIDC secretNamespace to locate client secrets.
  * Certificate-aware secret predicate to trigger reconciliations.
  * Auth-proxy manifests converted to templates with configurable component labels.

* **Improvements**
  * Hardened auth-proxy security (reduced replicas, stricter securityContext), adjusted cookie expiry/refresh and auth timeouts.
  * Centralized certificate/OAuth handling, streamlined reconciler and template-driven resource provisioning.
  * NetworkPolicy enabled by default and updated port/label semantics.

* **Documentation**
  * README and API docs updated for secretNamespace and NetworkPolicy defaults.

* **Tests**
  * Gateway test suite restructured; many legacy runtime-style tests removed and focused unit tests added.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: internal/controller/services/gateway/gateway_controller.go [hunk 1]
```diff
@@ -22,97 +22,59 @@ import (
...
-			reconciler.WithEventHandler(handlers.ToNamed(serviceApi.GatewayConfigName)),
-			reconciler.WithPredicates(predicate.GenerationChangedPredicate{}),
-		)
-
-	// Watch ingress certificate secrets to trigger reconciliation when certificates are rotated
-	// This ensures gateway certificates are automatically updated when the source certificate changes
-	reconcilerBuilder = reconcilerBuilder.
+	gw.OwnsGVK(gvk.GatewayClass).
+		OwnsGVK(gvk.KubernetesGateway).
+		// OwnsGVK(gvk.Secret).
```
Comment (davidebianchi) [inline internal/controller/services/gateway/gateway_controller.go]: Why secret now is commented out?

File: internal/controller/services/gateway/gateway_controller_actions_test.go
Comment (davidebianchi) [inline internal/controller/services/gateway/gateway_controller_actions_test.go:1]: Why all the unit tests removed?

File: internal/controller/services/gateway/gateway_support_test.go [hunk 2]
```diff
@@ -135,856 +78,548 @@ func TestGetCertificateType(t *testing.T) {
...
+		},
 	}
+	hash5 := calculateAuthConfigHash(secret5)
+	g.Expect(hash5).To(Equal(hash1), "same secret values should produce same hash")
 }
 
-// TestGetFQDNWithSubdomain tests subdomain functionality in domain resolution.
-func TestGetFQDNWithSubdomain(t *testing.T) {
+// TestCalculateAuthConfigHash tests the calculateAuthConfigHash function.
+func TestCalculateAuthConfigHash(t *testing.T) {
```
Comment (davidebianchi) [inline internal/controller/services/gateway/gateway_support_test.go]: This isn't already checked in TestSecretHashAnnotationChangeTriggers?

File: internal/controller/services/gateway/gateway_support_test.go [hunk 3]
```diff
@@ -135,856 +78,548 @@ func TestGetCertificateType(t *testing.T) {
...
+			result := isGatewayReady(tc.gateway)
+			g.Expect(result).To(Equal(tc.expectReady), tc.description)
 		})
 	}
 }
 
-// TestCheckAuthModeNone tests the None authentication mode validation logic.
-func TestCheckAuthModeNone(t *testing.T) {
+// TestGetCertificateTypeEdgeCases tests edge cases for certificate type detection.
+func TestGetCertificateTypeEdgeCases(t *testing.T) {
```
Comment (davidebianchi) [inline internal/controller/services/gateway/gateway_support_test.go]: Could be added to TestGetCertificateType?

File: internal/controller/services/gateway/gateway_support_test.go [hunk 4]
```diff
@@ -135,856 +78,548 @@ func TestGetCertificateType(t *testing.T) {
...
+			result := getCertificateType(tc.gatewayConfig)
+			g.Expect(result).To(Equal(tc.expectedType), tc.description)
 		})
 	}
 }
 
-// TestCreateKubeAuthProxyService tests the auth proxy service creation logic.
-func TestCreateKubeAuthProxyService(t *testing.T) {
+// TestCalculateAuthConfigHashDeterministic verifies hash function is deterministic.
+func TestCalculateAuthConfigHashDeterministic(t *testing.T) {
```
Comment (davidebianchi) [inline internal/controller/services/gateway/gateway_support_test.go]: Also this, isn't already checked in TestSecretHashAnnotationChangeTriggers?

File: internal/controller/services/gateway/gateway_support_test.go [hunk 5]
```diff
@@ -135,856 +78,548 @@ func TestGetCertificateType(t *testing.T) {
...
-	g.Expect(args).NotTo(ContainElement(ContainSubstring("--oidc-issuer-url=")))
+	g.Expect(hash1).To(Equal(hash2), "hash should be deterministic")
+	g.Expect(hash2).To(Equal(hash3), "hash should be deterministic")
+	g.Expect(hash1).To(HaveLen(64), "hash should be 64 characters (SHA256 hex)")
 }
 
-// TestBuildOAuth2ProxyArgsOIDC tests OAuth2 proxy arguments for OIDC mode.
-func TestBuildOAuth2ProxyArgsOIDC(t *testing.T) {
+// TestGetGatewayAuthProxyTimeoutEdgeCases tests edge cases for timeout retrieval.
+func TestGetGatewayAuthProxyTimeoutEdgeCases(t *testing.T) {
```
Comment (davidebianchi) [inline internal/controller/services/gateway/gateway_support_test.go]: Can we merge those test with TestGetGatewayAuthProxyTimeout?

File: pkg/cluster/cert.go [hunk 6]
```diff
@@ -188,6 +190,48 @@ func GetDefaultIngressCertSecretName(ingressCtrl *operatorv1.IngressController)
...
+
+	case infrav1.Provided:
+		expectedName := certConfig.SecretName
+		if expectedName == "" {
+			expectedName = fmt.Sprintf("%s-tls", gatewayConfig.Name)
+		}
+		return obj.GetName() == expectedName
+
+	default:
+		// no need actoin on selfsigned as operator create it which has the ownerreference on it with reconcile.
```
Comment (davidebianchi) [inline pkg/cluster/cert.go]:
```suggestion
		// no need action on selfsigned as operator create it which has the ownerreference on it with reconcile.
```

File: tests/e2e/gateway_test.go [hunk 7]
```diff
@@ -193,14 +199,14 @@ func (tc *GatewayTestCtx) ValidateOAuthClientAndSecret(t *testing.T) {
 		}),
 		WithCondition(And(
 			jq.Match(`.type == "%s"`, string(corev1.SecretTypeOpaque)),
-			jq.Match(`.metadata.labels.app == "%s"`, kubeAuthProxyName),
+			jq.Match(`.metadata.labels["platform.opendatahub.io/part-of"] == "gatewayconfig"`),
```
Comment (davidebianchi) [inline tests/e2e/gateway_test.go]: Should we use a constant for the label?

File: internal/controller/services/gateway/gateway_controller.go (hunk 1 above)
Comment (zdtsw) [reply internal/controller/services/gateway/gateway_controller.go]: you are right, we need own the secret created by gateway, the below watches is a different one.

File: internal/controller/services/gateway/gateway_controller_actions_test.go
Comment (zdtsw) [reply internal/controller/services/gateway/gateway_controller_actions_test.go:1]: the old unit-test are for the old way to create resources dynamically, since they are now moved to use template, they are not valid, the still needed one are moved into the `gateway_support_test.go`

File: internal/controller/services/gateway/gateway_controller_actions.go [hunk 8]
```diff
@@ -58,52 +59,246 @@ func createGatewayInfrastructure(ctx context.Context, rr *odhtypes.Reconciliatio
...
+	switch authMode {
+	case cluster.AuthModeOIDC:
+		if gatewayConfig.Spec.OIDC == nil {
+			rr.Conditions.MarkFalse(
+				ReadyConditionType,
+				conditions.WithReason(status.NotReadyReason),
+				conditions.WithMessage(status.AuthProxyOIDCModeWithoutConfigMessage),
+			)
+			return nil // TODO: is this logic correct? no oidc but to use oidc and not error out.
+		}
```
[the remark this answers is not in the document]
Comment (ugiordan) [reply internal/controller/services/gateway/gateway_controller_actions.go:119]: @GowthamShanmugam, what scenario allows OIDC mode without OIDC config? Is this a valid edge case or a bug?

File: internal/controller/services/gateway/resources/kube-auth-proxy-oidc-deployment.tmpl.yaml [hunk 9]
```diff
@@ -14,6 +15,7 @@ spec:
     metadata:
       labels:
         app: {{.KubeAuthProxyServiceName}}
+        {{.ComponentLabelKey}}: {{.ComponentLabelValue}}
       annotations:
```
Comment (GowthamShanmugam) [inline internal/controller/services/gateway/resources/kube-auth-proxy-oidc-deployment.tmpl.yaml:19]: internal/controller/services/gateway/resources/kube-auth-proxy-oauth-deployment.tmpl.yaml and internal/controller/services/gateway/resources/kube-auth-proxy-oidc-deployment.tmpl.yaml looks same

Comment (zdtsw) [reply internal/controller/services/gateway/resources/kube-auth-proxy-oidc-deployment.tmpl.yaml:19]:
```
diff kube-auth-proxy-oauth-deployment.tmpl.yaml kube-auth-proxy-oidc-deployment.tmpl.yaml
102,103c102,105
<             - "--provider=openshift"
<             - "--scope=user:full"
\ No newline at end of file
---
>             - "--provider=oidc"
>             - "--oidc-issuer-url={{.OIDCIssuerURL}}"
>             - "--skip-oidc-discovery=false"
>             - "--ssl-insecure-skip-verify=false" # TODO: need to check if this is correct to hardcode skip verify.
```
is this the diff for 2 kinds? or i missed something

File: internal/controller/services/gateway/gateway_support_test.go (hunk 4 above)
Comment (zdtsw) [reply internal/controller/services/gateway/gateway_support_test.go]: yep, after rename the fuction, only one test is needed, let me remove this

File: internal/controller/services/gateway/gateway_controller.go (hunk 1 above)
Comment (ugiordan) [reply internal/controller/services/gateway/gateway_controller.go]: We either uncomment this or extend the secret watch predicate to include:
- kube-auth-proxy-creds (all events)
- External OIDC secrets referenced in GatewayConfig (update events)

File: internal/controller/services/gateway/resources/kube-auth-proxy-oidc-deployment.tmpl.yaml (hunk 9 above)
Comment (GowthamShanmugam) [reply internal/controller/services/gateway/resources/kube-auth-proxy-oidc-deployment.tmpl.yaml:19]: ack

File: internal/controller/services/gateway/gateway_controller_actions.go [hunk 10]
```diff
@@ -58,52 +59,246 @@ func createGatewayInfrastructure(ctx context.Context, rr *odhtypes.Reconciliatio
...
+	var oidcConfig *serviceApi.OIDCConfig
+	switch authMode {
+	case cluster.AuthModeOIDC:
+		if gatewayConfig.Spec.OIDC == nil {
+			rr.Conditions.MarkFalse(
+				ReadyConditionType,
+				conditions.WithReason(status.NotReadyReason),
+				conditions.WithMessage(status.AuthProxyOIDCModeWithoutConfigMessage),
+			)
+			return nil // TODO: is this logic correct? no oidc but to use oidc and not error out.
```
Comment (ugiordan) [inline internal/controller/services/gateway/gateway_controller_actions.go]:
```suggestion
          // Return nil (not error) because this is a permanent user configuration error.
          // Returning an error would cause infinite reconciliation for something only the
          // user can fix. Users must check GatewayConfig status conditions to see this error.
          // Reconciliation will retry when the user updates the GatewayConfig with OIDC config.
          return nil
```

We have the same pattern in the monitoring controller where it does the same thing when metrics/traces aren't configured.

However, we should make this even better in future PRs:
- Admission webhook: It would be nice to validate the OIDC config at admission time so users get immediate feedback instead of having to check status conditions. 
- Watch auth mode changes: Right now, if someone switches the cluster from OAuth to OIDC, we won't know until something else triggers reconciliation. Could watch the Authentication CR if you want to be more responsive.
- Alerting: Maybe add a PrometheusRule if the gateway sits in NotReady for too long?

@GowthamShanmugam, @zdtsw WDYT? We can create follow-up Jiras.

File: internal/controller/services/gateway/gateway_controller_actions.go [hunk 11]
```diff
@@ -58,52 +59,246 @@ func createGatewayInfrastructure(ctx context.Context, rr *odhtypes.Reconciliatio
...
+			FS:   gatewayResources,
+			Path: kubeAuthProxyDeploymentOidcTemplate,
+		}
+	case cluster.AuthModeIntegratedOAuth: // default mode.
+		l.V(1).Info("configuring " + KubeAuthProxyName + " for OpenShift OAuth")
+
+	case cluster.AuthModeNone:
+		rr.Conditions.MarkTrue(
+			ReadyConditionType,
+			conditions.WithReason(status.ReadyReason), // TODO: is this logic correct? user do not want it, we should mark it as ready.
```
Comment (ugiordan) [inline internal/controller/services/gateway/gateway_controller_actions.go]: This is semantically correct. When cluster uses external auth, the desired state is "do not deploy auth proxy", so marking `Ready=True` is appropriate.

```suggestion
			// Ready=True is correct: desired state is "no auth proxy" when using external auth.
			// The status message clarifies: "Cluster uses external authentication, no gateway auth proxy deployed"
  			conditions.WithReason(status.ReadyReason), 
```

**Review by ugiordan — CHANGES_REQUESTED** [review]

File: internal/controller/services/gateway/gateway_controller.go (hunk 1 above)
Comment (zdtsw) [reply internal/controller/services/gateway/gateway_controller.go]: i uncommented it, to just keep what it was.

**Comment by ugiordan** [comment]
Just two minor things on the TODO comments. After those small clarifications, this is good to merge.

File: internal/controller/services/gateway/gateway_controller_actions.go (hunk 10 above)
Comment (zdtsw) [reply internal/controller/services/gateway/gateway_controller_actions.go]: sure, a better inline comments if can explain the "why for now" and follow up later, is much better than my "self asking questions"

**Review by ugiordan — APPROVED** [review]
